core/src/solve.jl

```diff
@@ -199,6 +249,83 @@ function formulate!(
     return nothing
 end
 
+
+"""
+Connection based on preservation of energy:
+
+h_a + v_a^2 / (2 * g) = h_b + v_b^2 / (2 * g) + h_loss
+
+With:
+
+h_loss = S_f * L + C / 2 * g * (v_b^2 - v_a^2)
+
+Where friction losses are approximated by the Gauckler-Manning formula:
+
+V = (1 / n) * R_h^(2/3) * S_f^(1/2)
+
+Where:
+
+    * V is the cross-sectional average velocity.
+    * n is the Gauckler-Manning coefficient.
+    * R_h is the hydraulic radius.
+    * S_f is the friction slope.
+    
+R_h = A / P
+
+    * Where A is the cross-sectional area.
+    * P is the wetted perimeter.
+
+"""
+function formulate!(
+    basin::Basin,
+    connectivity::Connectivity,
+    manning_connection::ManningConnection,
+    level,
+)::Nothing
+    (; graph, flow, u_index) = connectivity
+    (; node_id, length, n_squared, profile_width, profile_slope, unit_length, contraction_c, expansion_c) = manning_connection
+    for (i, id) in enumerate(node_id)
+        basin_1_id = only(inneighbors(graph, id))
+        basin_2_id = only(outneighbors(graph, id))
+
+        h_1 = level[u_index[basin_1_id]]
+        h_2 = level[u_index[basin_2_id]]
+        bottom_1 = first(basin.level[basin_1_id].u)
+        bottom_2 = first(basin.level[basin_1_id].u)
```

Comment:
The state `u` and the vectors in Basin are aligned, same length and order. The ID cannot be used as an index, so here you also need to use `u_index[basin_1_id]`, so it's probably a good idea to save that as a variable. Also note that you use `basin_1_id` in the second line instead of `basin_2_id`.